libacb/util/deps.go

```diff
@@ -0,0 +1,354 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"crypto/tls"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/appc/spec/discovery"
+	"github.com/appc/spec/schema/types"
+	"github.com/coreos/ioprogress"
+)
+
+var (
+	defaultArch = "amd64"
+	defaultOS   = "linux"
+)
+
+// RenderACI will download any missing dependencies into the depstore,
+// expand any unexpanded dependencies into the scratchpath, and return
+// an ordered list of all dependencies for overlaying.
+func RenderACI(acipath, scratchpath, depstore string, insecure bool) ([]string, error) {
+	man, err := GetManifest(acipath)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(man.Dependencies) == 0 {
+		return nil, nil
+	}
+
+	remainingPaths := []string{acipath}
+	index := 0
+
+	for len(remainingPaths) > index {
+		depPaths, err := renderACI(remainingPaths[index], scratchpath, depstore, insecure)
+		if err != nil {
+			return nil, err
+		}
+
+		tmpList := append(remainingPaths[:index+1], depPaths...)
+		remainingPaths = append(tmpList, remainingPaths[index+1:]...)
+		index++
+	}
+
+	return remainingPaths[1:], nil
+}
+
+func renderACI(acipath, scratchpath, depstore string, insecure bool) ([]string, error) {
+	man, err := GetManifest(acipath)
+	if err != nil {
+		return nil, err
+	}
+
+	err = FetchDeps(man.Dependencies, depstore, insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	var expandedDeps []string
+	for _, dep := range man.Dependencies {
+		depname := string(dep.ImageName)
+		expandedpath := GenDepPath(scratchpath, depname)
+
+		ex, err := Exists(expandedpath)
+		if err != nil {
+			return nil, err
+		}
+		if ex {
+			expandedDeps = append(expandedDeps, expandedpath)
+			continue
+		}
+
+		err = os.Mkdir(expandedpath, 0755)
+		if err != nil {
+			return nil, err
+		}
+
+		err = ExpandACI(GenDepPath(depstore, depname), expandedpath)
+		if err != nil {
+			os.RemoveAll(expandedpath)
+			return nil, err
+		}
+
+		expandedDeps = append(expandedDeps, expandedpath)
+	}
+
+	reversedDeps := make([]string, len(expandedDeps))
+	for i, dep := range expandedDeps {
+		reversedDeps[len(expandedDeps)-i-1] = dep
+	}
+
+	return reversedDeps, nil
+}
+
+func ExpandACI(acipath, targetpath string) error {
+	//TODO: there should be a way to avoid exec'ing another binary here
+	cmd := exec.Command("tar", "xf", acipath, "-C", targetpath)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+//TODO: path should include labels probably
+func GenDepPath(depstore, imageName string) string {
+	return path.Join(depstore, strings.Replace(imageName, "/", "-", -1)+".aci")
+}
+
+func FetchDeps(deps types.Dependencies, depstore string, insecure bool) error {
+	for _, dep := range deps {
```

Comment:
In future I'll move this the rkt fetching logic and avoid duplicating some work bugfixes (it should be moved, as already talked somewhere, to its own package)